### ISSUE TITLE
v0.13.2.AP: Pre-Release REL build does not display boot screen on MacPro5,1 with NVIDIA GTX 760

### ISSUE THREAD

**Issue by veeceeoh** [dakanji/RefindPlus#104]
### RefindPlus Version

Another Released Version

### Device Type

Apple Mac

### Problem Description

I am having an issue with the _release_ .EFI binary of RefindPlus **v0.13.2.AP Pre-Release** (included with MyBootMgr v079) not displaying the graphical boot selection menu. Note that the _debug_ binary of RefindPlus v0.13.2.AP Pre-Release displays the menu just fine.

The computer on which I am experiencing the issue is a mid-2010 Mac Pro (_MacPro5,1_) "Eight Core" (Dual 2.4 GHz Quad Core Xeon E5620 processors), and the video card installed in this Mac Pro is an **EVGA NVIDIA GTX 760**.

The RefindPlus `config.conf` was built using MyBootMgr v079's ConfigFactory app. In that `config.conf` file, the following relevant configuration items are set to _active_:
`provide_console_gop`
`text_renderer`
`uga_pass_through`

Attached is the RefindPlus `config.conf` file used on the Mac Pro having this issue, and the Debug Log from ConfigFactory, in a .zip file: [RFP_13_2_AP_Display_Issue.zip](https://github.com/dakanji/RefindPlus/files/8399142/RFP_13_2_AP_Display_Issue.zip)

### Problem Point

Before starting a loader or tool

### Affected Items

Other Item

### Debug Log

[22f15s0848.log](https://github.com/dakanji/RefindPlus/files/8399150/22f15s0848.log)

### Additional Context

Note that the _release_ .EFI binary of RefindPlus v0.13.2.**AN Alt** also does not display the boot menu on the affected mid-2010 Mac Pro when used with the above-attached `config.conf` file, while the _debug_ binary does display the menu. However, neither the release nor debug binaries of the non-Alt RefindPlus v0.13.2.**AN** release displays the RefindPlus boot menu.

**Comment by dakanji** [comment]
Just to get a better picture, when you blindly select a loader from that REL file, such as OpenCore, do you get a screen there?

**Comment by veeceeoh** [comment]
Yes, the OpenCore boot selection screen, native macOS load screen with the grey Apple logo, and Windows load screen do display as expected.

**Comment by dakanji** [comment]
Okay ... limited to RefindPlus.

Please try the attached build of the current code, X222, as the build you used has had a few changes in the intervening period since.

[X222-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8399523/X222-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _release_ binary of the X222 build still does **not** display the RefindPlus boot screen. 

The _debug_ X222 binary does work as expected.

**Comment by dakanji** [comment]
Thanks. Need to try to figure it out.

I believe you have access to another similar MacPro ... Have you run the same file on that?
If it works on that one, can you try swapping the GPUs to see whether the issue follows the GPU?

In the meantime, can you share the debug log from this run of X222?

**Comment by veeceeoh** [comment]
Apologies for the delay. Here is the debug log output from the X222 _debug_ binary: 
[22h05a3037.log](https://github.com/dakanji/RefindPlus/files/8415080/22h05a3037.log)

I'm not sure when I will have an opportunity to try swapping the NVIDIA GTX 760 videocard into my other 6-core MacPro5,1 because now I am experiencing a **more serious issue** with RefindPlus & OpenCore:

When using the "OpenCore (Default)" selection of a MyBootMgr v079 built ESP (I reverted back from your X222 binary to the original RefindPlus 0.13.2.AP Pre-Release debug binary included with  and then selecting any macOS Big Sur 11.6.4 installation, the 8-core Mac Pro is rebooting due to a kernel panic just some seconds after the Apple logo boot screen with progress bar appears. 

Since this kernel panic happens at boot time and the kernel panic dump log is only presented after macOS has successfully booted, I tried booting in Verbose mode while taking photos of the screen. I saw this message:

_"Rooting from the live fs of a sealed volume is not allowed on a RELEASE build"_

Upon some Googling, I saw an older post on the MacRumors Forums' MyBootMgr thread where user Macschrauber said he was seeing the above error when chainloading RefindPlus > OC > macOS Big Sur, and was only able to get it to boot if SIP is disabled. 

I did not have any issue booting with a MyBootMgr v079 ESP build on this Mac Pro 8-core machine until I tried the X222 binary, and now even though I have reverted back to the 0.13.2.AP Pre-Release debug binary I am still seeing a Kernel Panic reboot every time. It is also wreaking havoc on the Mac's NVRAM, so I have tried flashing the BootROM with a reconstructed one I received from MacRumors Forums user tsialex. When I tried booting into macOS Big Sur after that (again chainloaded via RefindPlus -> OpenCore (Default selection)), the Mac rebooted with a kernel panic yet again, so fully clearing the NVRAM did not resolve the issue. Again, I was not having these Kernel Panics happen previously. Booting natively into macOS Mojave works fine without issue, and if I _disable_ SIP in the RefindPlus boot menu, then I can boot into macOS Big Sur via the OC (Default) selection built by MyBootMgr. I'd really rather not have SIP disabled, however.

I am not sure what steps to take. Is there any way you can help? Please let me know if you'd like me to write this all out again over on your MacRumors Forums thread.

**Comment by dakanji** [comment]
Right. A long post coming up as some things need to be cleared up in order to proceed...

**==LONG STUFF: START==**
While it seems you have one big muddled up issue, there are actually two separate issues: RefindPlus Issue and OpenCore Issue.

RefindPlus does not affect OpenCore and any issue you find in OpenCore is an OpenCore issue reagrdless of whether RefindPlus is in play or not. There are some caveats, but you need to see both as totally separate and independent programs. 

The understandable difficult with grasping this has been a thorn in my side since MyBootMgr started as it is understandable for someone to conclude that they have one big marshup called RefindPlus/OpenCore and that since RefindPlus appears before OpenCore, issues seen in OpenCore are likely related to RefindPlus but the reality is that both are independent and operate separately. All MyBootMgr does is to configure both to work together.

A second consideration is that Concurrence does not mean Causation. That is, the fact that two things happen together or in sequence, does not mean the one has led to the other. Concurrence is understandably often linked to Causation but while Concurrence can imply Causation, it is not a given.

The thing to do is to carry out elimination tests as you have done:

- You changed to another binary
- You flashed in a clean NVRAM since the issue could have been persisted in the NVRAM

What those two tests say is that the binary test was just pure coincidence and something else must have changed. However, in my time dabbling with software stuff, it is often the case that the belief that Concurrence does infact mean Causation is so strong that tests that say otherwise are typically discounted.

Basically, there is a tendency to have decided on what is wrong, and any "troubleshooting" done is often subconsciously actually mainly to confirm this position as opposed to determining what is actually the problem. If the troubleshooting does not conform, it tends to be discarded.

Anyway, I want to put the OpenCore issue to one side and will get back to this later as it is in fact not a serious issue.
That is, from what I can gather, your issues only exist because you have decided to manually modify the MyBootMgr output config file and i**t goes away once you revert to the output config**. So stick to the output config for now and we can look at this later.
**==LONG STUFF: END==**

Now, all that long stuff is to get to this point:
For now, the more important item is the RefindPlus issue as no one else seems to be able to replicate the issue you raised about the REL file. So, please try X224 which is a REL file: 
[X224-BOOTx64-REL.zip](https://github.com/dakanji/RefindPlus/files/8416112/X224-BOOTx64-REL.zip)

All the long stuff was just to make you hopefully see that the RP binary, such as X222 or X224, have nothing to do with the OpenCore issues and are perfectly safe to boot.

You can of course just boot into MacOS directly from RP afterwards if you prefer, or shut down and revert to something you are more comfortable with. I just need to know whether the REL file shows the screen, or not, to move forward on the RP issue. The OC issue can be picked up later as said.

Thanks.

**Comment by veeceeoh** [comment]
Thanks for the reply; I appreciate your help. 

I tried the X224 release binary, and the RefindPlus menu **still does not display**. I was able to select a boot option "blindly" by pressing the corresponding number on my keyboard, however, and after a wait of approximately 10 seconds the macOS boot screen with Apple logo and progress bar was displayed. In this case, I had selected to boot macOS Mojave "natively" (directly from RefindPlus, not by chain-loading OpenCore).

Regarding the separate Kernel Panic issue, I'd also like to clarify a few things in response to your long reply:

<details>
 <summary>Clarifications</summary> 

- I am well aware that RefindPlus and OpenCore are two completely separate software solutions. I was using rEFInd (the Roderick W. Smith fork) to help with dual-booting between macOS Mojave and Windows 10 before I gave MyBootMgr a try as a way to also boot into macOS Big Sur as well. I reviewed the possibility of using OpenCore solely, and decided that the chain-loading solution offered by MyBootMgr was more elegant and much easier to manage than with OpenCore alone.
- I also do not see the two issues as connected. I was not sure where is the best place to ask you about it so I explained this new issue in this thread, and did ask whether you preferred that I ask about it on the MyBootMgr MacRumors Forums thread.
- Despite RefindPlus and OpenCore being two separate software solutions, I see that there are aspects of booting and running macOS that they both can have an effect on. The `csr-active-config` value setting is an example of this. Both solutions can make changes to it, so it is completely possible for a RefindPlus configuration to have an influence on macOS booted via OpenCore when chain-loaded from RefindPlus.
- When troubleshooting, I **never assume concurrence equals causation**. I merely wanted to note that I was not having the Kernel Panic at macOS Big Sur boot time issue before switching to the RefindPlus 0.13.2.AP X222 Pre-Release binary. I feel it is important to note any change(s) made just before an issue started, not to assume the change is the cause, but rather that it should be considered as a _possible_ cause, to be ruled out.
- Another troubleshooting practice I follow is to **revert to the last-known working state**. So I reverted back to the RefindPlus 0.13.2.AP Pre-Release debug binary supplied by the MyBootMgr v079 build, which is what was in place before the Kernel Panic at at macOS Big Sur boot time began. I also reverted the MacPro5,1's NVRAM to a known working state.
- I will only ask for advice or help on the RefindPlus and OpenCore configurations built by MyBootMgr when I have observed an issue with a **completely unmodified MyBootMgr build**, because any customizations made to any of the configuration files are outside the realm of what the MyBootMgr solution provides, and would be wasting time. In this case, I made sure that I was using a completely untouched EFI folder as built by MyBootMgr v079 before writing about the Kernel Panic issue. 
</details>

**Comment by dakanji** [comment]
Okay thanks.

The Refindplus REL thing not showing up is a wierd one but unfortunately, I don't know where to start with that.

The only way forward is to keep going back down the RefindPlus versions until you get to a working one. We can then try to find a breaking point. It will be a bit of a slog but if you are up for it, I will work with you on that.

You have pointed out that v0.13.2.AN does not display for you, so please try v0.13.2.AM and work back until you get to one that does. We can then try to find where it breaks: https://github.com/dakanji/RefindPlus/releases

Let me know if you are up for this.

Thanks


***


Thanks for the clear stuff on the MyBootMgr stuff. Very well written and enlightening.

Let's move that one over to the MacRumors MyBootMgr thread. I think the best way will be to post the text in the drop down here:  https://github.com/dakanji/RefindPlus/issues/104#issuecomment-1088280213 as a new post.

Would be helpful if you could zip up your EFI folder (Entire RP/OC) and attach to that post or provide a link to it somewhere if too large.

**Comment by veeceeoh** [comment]
So, the most recent RefindPlus _release_ binary that displays the RefindPlus boot menu on my MacPro5,1 8-core w/GTX 760 video card is v0.13.2.AM (from download asset named _RefindPlus_x64-001302_AM-3.zip_, dated December 25, 2021).

Here is the debug log output from the RefindPlus v0.13.2.AM _debug_ binary: [22h06t2239.log](https://github.com/dakanji/RefindPlus/files/8429175/22h06t2239.log)

It may be a while before I can try swapping the GTX 760 graphics card to my MacPro5,1 6-core machine because I am using it for working remotely.

**Comment by dakanji** [comment]
> The most recent RefindPlus REL binary that displays the RefindPlus boot menu is v0.13.2.AM

That's good, as the items addressed in the updates can be replicated in the current codebase. So a bit easier to handle a search for the break point.

Will start prepping the test files. Might take up to 20 rounds to narrow the issue down unless we get lucky.

> It may be a while before I can try swapping the GTX 760 graphics card

No need for that actually.

**Comment by dakanji** [comment]
Please try R000 for a baseline.
[R000-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8430561/R000-BOOTx64.zip)

**Comment by dakanji** [comment]
Just looked at the R000 file and noticed I had deleted the `.efi` extension while naming it. This might be why it failed. Can you add this and try again.

On availability, hours away are fine and expected. You can even go to bed now and then!
Just pick up on it when you are available and hopefully not needing a day in between. that all.

On test setup, up to you if what you described works for you but I test a shed load of builds and can't afford to dismantle/disconnect stuff each time, so I can tell you that the ESP on a USB approach is the best way to go about it.

Whatever works for you in the end.

Anyway, take a look at the extension thing when you have a minute.

**Comment by veeceeoh** [comment]
For all of these tests, I am renaming the test binary to BOOTx64.efi and replacing the one in the EFI\BOOT folder. I know this works by checking the version on the know working binaries in the About window.

I did this for the R000 test file.

**Comment by dakanji** [comment]
Okay. Let's see how R000a does in that case.
[R000a-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8445253/R000a-BOOTx64.zip)

Made a small adjustment that might make a difference.
Still just trying to get a baseline before starting the search

**Comment by veeceeoh** [comment]
The _R000a_ doesn't display the RefindPlus boot menu, but must be loading because I see activity on the USB flash drive every time I press the escape key.

**Comment by dakanji** [comment]
Thanks.

Please try R000b
[R000b-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8445801/R000b-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _R000b_ binary doesn't display the RefindPlus boot menu, but is loading.

**Comment by dakanji** [comment]
Thanks. Was hoping we could find a base to work from in the current code as applying a fix would have been easier but seems it is not to be. Now targeting the state close to 0.13.2.AM in R001.
[R001-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8446076/R001-BOOTx64.zip)

**Comment by veeceeoh** [comment]
Success! The _R001_ binary displays the RefindPlus boot menu. The version listed in the _About..._ screen is v0.13.2.AN (Pre-Release).

**Comment by dakanji** [comment]
OK. We now need to move towards the current code in steps to find the breaking commit.

There are 24 commits between this and the v0.13.2.AN release, which you previously confirmed does not work. So the breaking commit must be one of those. 

Will do 8 steps each. If R002 fails, we drop back 4 commits. If it works, we move up by 8 commits. 
[R002-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8446184/R002-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _R002_ binary displays the RefindPlus boot menu.

**Comment by dakanji** [comment]
Okay. Try R003.
[R003-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8446279/R003-BOOTx64.zip)

We will go up or down by 4 depending on the outcome

**Comment by veeceeoh** [comment]
The _R003_ binary does not display the RefindPlus boot menu, and it does not appear to be loading (no activity light on pressing escape - and I tried booting 4 times to be sure.)

**Comment by dakanji** [comment]
Thanks ... sorry for disappearing without notice.

R003 works for me but let's put it as the first known bad instance. while R002 stays as the last known good one.

Please try R004
[R004-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8448848/R004-BOOTx64.zip)

**Comment by veeceeoh** [comment]
It’s bedtime in my time zone, so I’ll be able to pick up with testing in around 9.5 hours. See you then!

**Comment by veeceeoh** [comment]
The _R004_ binary displays the RefindPlus menu.

**Comment by dakanji** [comment]
Thanks. Will post another in about an hour.

**Comment by dakanji** [comment]
Okay ... Please try R005
[R005-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8453823/R005-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _R004_ binary displays the RefindPlus menu.

**Comment by dakanji** [comment]
The breaking commit is the [R003 Commit](https://github.com/dakanji/RefindPlus/commit/65e891f11f47aafb0bdf60faebb7691c5e4f4b7b)

It was a pretty basic change done there but there is a likelihood that the changes exposed some other deeper lying issue that just happened to manifest for you for some other quirk of your setup and is still there, albeit not manisfested, on other units.

 Anyway, try X225. R003 change reverted in the current code.
[X225-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8454132/X225-BOOTx64.zip)

If it works, I'll try to find another way to achieve the same goal.

**Comment by veeceeoh** [comment]
The _X225_ binary does not display the RefindPlus menu, but is loading.

**Comment by dakanji** [comment]
Not surprised. Basically confirms the changes themselves were not the real issue.
Anyway, let's see whether reverting them in the 0.13.2.AN release makes a difference.
[R006-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8454322/R006-BOOTx64.zip)

This one might hang btw

**Comment by veeceeoh** [comment]
The R006 binary appears to hang, and does not display the RefindPlus menu.

**Comment by dakanji** [comment]
Okay, lets see with R007
[R007-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8454680/R007-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _R007_ binary does not display the RefindPlus menu, but does load.

**Comment by dakanji** [comment]
Seems we hit a bit of a dead end on this one.

Needs thinking about. Let's take a break till tomorrow.

**Comment by dakanji** [comment]
Can you try X227? It's a bit of a stab in the dark to rule an old issue out.
[X227-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8455737/X227-BOOTx64.zip)

If it doesn't work, then its the thinking hat back on

**Comment by veeceeoh** [comment]
The X227 binary loads, but doesn't display the RefindPlus menu.

**Comment by dakanji** [comment]
Hi ... Please try R006 when you have a min
[R006-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8456871/R006-BOOTx64.zip)

**Comment by veeceeoh** [comment]
Sorry for the delay; it’s busy for me this weekend. This new _R006_ binary works! The RefindPlus menu is displayed.

**Comment by dakanji** [comment]
OK. Try this new R007.
[R007-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8457692/R007-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The new _R007_ binary loads, but does not display the RefindPlus menu.

**Comment by dakanji** [comment]
Thanks.
Please try X228
[X228-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8457820/X228-BOOTx64.zip)

**Comment by veeceeoh** [comment]
The _X228_ binary works - it does display the RefindPlus menu.

**Comment by dakanji** [comment]
Let's leave things with that workaround for now.
Will decide on how best to proceed later.

Thanks for the time and effort

**Comment by veeceeoh** [comment]
Sure thing - thank you for the help and patience!

**Comment by dakanji** [comment]
The only other thing I can think of is that you test something like [X224 above](https://github.com/dakanji/RefindPlus/issues/104#issuecomment-1088395618) on your other machine.

If you have done the USB for chain-loading as suggested, you could just plug it in to that and test this easily without needing to make any changes to that machine (apart from an amendment to the custom.conf file to allow the chain-loading).

**Comment by veeceeoh** [comment]
Hi - again sorry for the delay! I checked and the _X224_ binary does load and display the boot screen on my other MacPro5,1 (a mid-2010 single 3.33GHz 6-core with XFX AMD Radeon RX 580 8GB video graphics card.)

**Comment by dakanji** [comment]
Okay ... Scrubbing the release as there are too many moving parts.

On this, looks to be something specific to this machine. Does not mean there is something wrong with the machine btw but the end result is an inability to nail down and fix the issue at this time.

Noting and suspending this for now

**Comment by dakanji** [comment]
Hi ... Can you please give X240 a try on this issue?
[X240-BOOTx64-REL.zip](https://github.com/dakanji/RefindPlus/files/8694351/X240-BOOTx64-REL.zip)

Someone else has mentioned something about REL behaving differently to DBG and I had actually come across a block of code that could affect display in some instances which was wrapped in DEBUG tags meaning it was only in the DBG build but should have been in both DBG and REL.

**Comment by veeceeoh** [comment]
Hi, I am sorry to say that I won’t be able to do any more testing of this issue because I just finally sold the MacPro5,1 "Eight Core" with NVIDIA GTX 760 graphics card.

**Comment by dakanji** [comment]
I see. Thanks for getting back.
I will tag this as a known issue for future reference in that case.
